### ISSUE TITLE
Bug in handling large parameter values

### ISSUE THREAD

**Issue by martinholmer** [open-source-economics/Tax-Calculator#589]
The following problem was encountered on the master branch using the 3a87f50 commit.
Consider the following test:
```
@pytest.mark.one
def test_high_mte():
    """
    Test different ways of eliminating the maximum taxable earnings (MTE)
    used in the calculation of the OASDI payroll tax.
    """
    # clarify start year and create Policy parameters object
    syr = 2013
    ppo = Policy(start_year=syr)
    # confirm that MTE (or '_SS_Earnings_c') has current-law value in 2016
    assert ppo._SS_Earnings_c[2016 - syr] == 118500
    # specify a "pop the cap" reform that 'eliminates' MTE cap in 2016
    # ... first way:
    reform = {2016: {'_SS_Earnings_c': [9.20E18]}}
    ppo.implement_reform(reform)
    assert ppo._SS_Earnings_c[2016 - syr] == 9200000000000000000
    # ... second way:
    reform = {2016: {'_SS_Earnings_c': [9.25E18]}}
    ppo.implement_reform(reform)
    assert ppo._SS_Earnings_c[2016 - syr] == 9250000000000000000
```

When I add this test to the test_policy.py file and run the test, I get these results:
```
$ py.test -m one
============================= test session starts ==============================
platform darwin -- Python 2.7.11, pytest-2.8.1, py-1.4.30, pluggy-0.3.1
rootdir: /Users/mrh/work/OSPC/tax-calculator, inifile: setup.cfg
plugins: pep8-1.0.6
collected 108 items 

tests/test_policy.py F

=================================== FAILURES ===================================
________________________________ test_high_mte _________________________________

    @pytest.mark.one
    def test_high_mte():
        """
        Test different ways of eliminating the maximum taxable earnings (MTE)
        used in the calculation of the OASDI payroll tax.
        """
        # clarify start year and create Policy parameters object
        syr = 2013
        ppo = Policy(start_year=syr)
        # confirm that MTE (or '_SS_Earnings_c') has current-law value in 2016
        assert ppo._SS_Earnings_c[2016 - syr] == 118500
        # specify a "pop the cap" reform that 'eliminates' MTE cap in 2016
        # ... first way:
        reform = {2016: {'_SS_Earnings_c': [9.20E18]}}
        ppo.implement_reform(reform)
        assert ppo._SS_Earnings_c[2016 - syr] == 9200000000000000000
        # ... second way:
        reform = {2016: {'_SS_Earnings_c': [9.25E18]}}
        ppo.implement_reform(reform)
>       assert ppo._SS_Earnings_c[2016 - syr] == 9250000000000000000
E       assert -9223372036854775808 == 9250000000000000000L

tests/test_policy.py:632: AssertionError
====================== 107 tests deselected by "-m 'one'" ======================
=================== 1 failed, 107 deselected in 3.70 seconds ===================
$ 
```

Notice that there was no Python error generated, so that if this was done in a production version of the Tax-Calculator, the maximum taxable earnings level for the OASDI payroll (FICA) tax would be set at a negative value, not at a very high value.   This kind of behavior has been seen when using TaxBrain (see the webapp-public repository's issue 138).  We can see from this test that the problem is not with TaxBrain or in the dropQ logic, but is inherent in the current logic of Tax-Calculator, which represents policy parameters as 64-bit integers.  Related to this integer representation of policy parameters, indexed parameters are always rounded down to the next lowest integer, which is not intuitive and has caused confusion among even the most experienced Tax-Calculator developers.

I think the current logic is unacceptable: we simply cannot have Tax-Calculator and TaxBrain saying that raising the maximum taxable earnings to a high level will **lower** FICA taxes, which is what they say now.

cc @MattHJensen @feenberg @talumbau @GoFroggyRun

**Comment by talumbau** [comment]
#590 handles this issue by changing the dtype of every column. It might be best for the dtype of the records to be `float64` rather than `int64`, but that's a separate issue from the bug specified here. The issue is that the user has specified a value for a parameter that overflows the dtype of the given record. A more modest change is #591 that simply raises an `OverflowError` when someone attempts to do this.

The explanation for the error behavior is as follows: 

The four types of interest are as follows:

 - Python integer: arbitrary precision integer type
 - Python float: IEEE 754 64-bit floating point number
 - numpy `int64`: 64-bit signed integer type
 - numpy `float64`: IEEE 754 64-bit floating point number

In the current master, we have this code in `parameters_base.py`:

```
cval[(year - self.start_year):] = nval
```
If nval has dtype `float64`, this triggers the 'broadcast' numpy behavior which goes directly fro `float64` to `int64`. Numpy does not check for overflow in this case, as described here on the numpy mailing list:

https://mail.scipy.org/pipermail/numpy-discussion/2009-April/041691.html

In #591, we copy value-by-value from `nval` to `cval` which takes each value through the following type transformation: `float64 -> Python float -> Python integer -> int64`. The last transition fails if there is an overflow because Python checks for overflow in that case.

I'm not directly opposed to #590, it just seems like the best way to hit the particular issue raised here is to properly catch and raise overflow errors.

**Comment by MattHJensen** [comment]
> I'm not directly opposed to #590, it just seems like the best way to hit the particular issue raised here is to properly catch and raise overflow errors.

@talumbau, I think one of the problems with this approach is that it would be setting up many TaxBrain users for overflow errors given our current TaxBrain input page interface. 

A major policy reform that some people like to consider is getting rid of the ceiling on earnings that are subject to the social security payroll tax. (http://www.npr.org/2013/12/05/249068448/to-fix-social-security-some-democrats-want-to-lift-wage-cap). 

The only way we allow TaxBrain users to do that right now is enter a really big number here

![image](https://cloud.githubusercontent.com/assets/8301092/12960284/7e7c8542-d008-11e5-81ff-678d3f2ae63c.png)

If they get overflow errors for really big numbers, they'll have to try to thread the needle between two numbers that they don't know up front 1) the minimum the number needs to be to approximate infinity and 2) the maximum the number can be to not cause an overflow.

**Comment by martinholmer** [comment]
Matt said:
>  T.J., I think one of the problems with this approach is that it would be setting up many TaxBrain users for overflow errors given our current TaxBrain input page interface.  ...[snip explanation of "pop-the-cap" FICA reforms]...

One advantage of pull request #590 is that Tax-Calculator and TaxBrain will be able to handle any large number (integer or float) for the maximum taxable earnings level (or any other parameter).  With the changes in pull request #590, I have successfully used ```1.0E+1000``` in a reform JSON file, and I have used ```float('inf')``` in a reform dictionary, for the post-pop-the-cap-reform parameter value.

**Comment by talumbau** [comment]
Ah I see. OK, then the only issue I would need to know about on the web app side is as follows: 

when the user inputs his own policy parameter choices, they will be interpreted as floating point values. Some of these values are advanced in time via some inflation rate. For these future years, would the new values be:

 - the complete floating point value of `user_val * (1.0 + inf_rate)`
 - the floating point value truncated to two decimal points of `user_val * (1.0 + inf_rate)`
 - the floating point value rounded to two decimal points of `user_val * (1.0 + inf_rate)`
 - something else?

**Comment by feenberg** [comment]
Is there a reason not to use floats or doubles? In taxsim I always use 
1e20 for infinity - it is a valid float or double and bigger than any 
number that could actually appear on a tax return. Since IEEE floats go up 
to 3.4x10**38 it is fairly easy "thread the needle" if they are used.

dan

**Comment by martinholmer** [comment]
Dan asked:
> Is there a reason not to use floats or doubles?

Pull request #590 moves in exactly that direction by representing parameters as float64 numbers.
In numpy, float64 is the same as double or REAL*8 in other languages.

There is no need to approximate infinity, because the IEEE float64 standard supports the number ```inf```, which can be specified in Python as ```float('inf')```.  In JSON specifications of a policy reform, one simply uses a very large number like 1e999, which gets converted to ```float('inf')``` in Python.

@feenberg @MattHJensen @talumbau @GoFroggyRun

**Comment by talumbau** [comment]
Sounds reasonable to handle everything with 8 byte floating point values. Just need to know if rounding/truncating should ever be imposed in certain intermediate calculations.

**Comment by martinholmer** [comment]
Dan asked:
> Is there a reason not to use floats or doubles?

Pull request #590 moves exactly in that direction by using float64 numbers for parameter values.
In numpy, a float64 number is the same as double or REAL*8 numbers in other languages.

If we move in this direction, there is not even the need to approximate infinity because the IEEE float64 standard specifies a number that represents infinity.  In a Python reform dictionary, infinity can be specified as ```float('inf')```.  And in a JSON reform file, infinity can be represented by a very large floating-point value (for example, 1E999), which is converted in Python to ```float('inf')```.

cc @feenberg @MattHJensen @talumbau @GoFroggyRun

**Comment by martinholmer** [comment]
T.J. asked:
> Sounds reasonable to handle everything with 8-byte floating point values [i.e., float64].  Just need to know if rounding/truncating should ever be imposed in certain intermediate calculations.

Do you mean "intermediate calculations" in Tax-Calculator?  If so, then I've tried to remove all those that I saw (but maybe I missed some).  I have no idea what you're doing in the TaxBrain code.  Why would there be a reason to round or truncate when doing "intermediate calculations" in the TaxBrain code?  Aren't all the user inputs just sent to Tax-Calculator to deal with?  Can you give an example of what prompted your question?

**Comment by talumbau** [comment]
Martin asked:
> Can you give an example of what prompted your question?

Sure. On TaxBrain, a user can specify a start year of tax year 2017. We then populate all of the parameter inputs with a default starting value. Many of these parameter values are not specified for the year 2017. Some are specified for 2016 and have metadata that says they should be inflated. So we find the 2017 value by inflating the 2016 value. We then display that value as a "greyed out" value in the TaxBrain input page. For example, take `_II_em`, with `long_name` value "Personal and dependent exemption amount." If I choose a start year of 2017 on TaxBrain, here is what I get:

<img width="455" alt="screen shot 2016-02-11 at 9 21 14 am" src="https://cloud.githubusercontent.com/assets/1891418/12980330/a7a0f734-d0a1-11e5-9fda-f631e005a603.png">

This is the result of a calculation that is something like this:

```
value_for_2017 = int(_II_em[2016] * (1.0 + cpi_inf[2016]))
```
(not the actual code but just for demonstration purposes)

As far as I can see, all specified values are given to the nearest dollar, so I'm just thinking it would be weird to populate that box with the full floating point value. So that is why I was wondering if the value that we should use in that case should be rounded to two decimal places or truncated to two decimal places or something like that.

**Comment by martinholmer** [comment]
@talumbau, Thanks for the clear example involving presentation of 2017 policy parameters in TaxBrain.  Your example raises a number of issues that I need to think about over the coming few days.

cc @MattHJensen @feenberg

**Comment by MattHJensen** [comment]
@talumbau said

> As far as I can see, all specified values are given to the nearest dollar, so I'm just thinking it would be weird to populate that box with the full floating point value. So that is why I was wondering if the value that we should use in that case should be rounded to two decimal places or truncated to two decimal places or something like that.

Let's round to the nearest dollar. @talumbau, would that be time consuming to implement on TaxBrain if we merge #590? 

As an aside, is there a way to achieve what #590 does without all of the tolerances business?

**Comment by MattHJensen** [comment]
@talumbau, I just merged #590. Once you are set with the rounding on TaxBrain, could you close this issue?

**Comment by martinholmer** [comment]
Matt said:
> @talumbau, I just merged pull request #590.  Once you are set with the rounding on TaxBrain, could you close this issue?

The webapp-public repository pull requests 159 and 161 would seem to implement the TaxBrain policy parameter rounding logic, so this tax-calculator repository issue #589 is being closed.
